<a href="https://colab.research.google.com/github/sujit-kumar-singh-9115/Alma-Better-Capstone-Project/blob/main/AlmaBetter_Hotel_Booking_Analysis_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## <b> Have you ever wondered when the best time of year to book a hotel room is? Or the optimal length of stay in order to get the best daily rate? What if you wanted to predict whether or not a hotel was likely to receive a disproportionately high number of special requests? This hotel booking dataset can help you explore those questions!

## <b>This data set contains booking information for a city hotel and a resort hotel, and includes information such as when the booking was made, length of stay, the number of adults, children, and/or babies, and the number of available parking spaces, among other things. All personally identifying information has been removed from the data. </b>

## <b> Explore and analyze the data to discover important factors that govern the bookings. </b>

In [78]:
# Importing packages
import pandas as pd
import numpy as np
from datetime import datetime
from datetime import date
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [79]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [80]:
# Reading the data file from the drive.
hotel_booking_df  = pd.read_csv('/content/drive/MyDrive/Alma-Better Folder/Module 1/Capstone project - EDA/Hotel Bookings.csv')

# Let's Ckeck the Data

In [81]:
# let's check the data.
hotel_booking_df

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
2                    July                        27   
3                    July                        27   
4                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
0                          0       2  ...    No Deposit    NaN     NaN   
1                          0       2  ...    No Deposit    NaN     NaN   
2                          1       1  ...    No Deposit    NaN     NaN   
3                          1       1  ...    No Deposit  304.0     NaN   
4                          2       2  ...    No Deposit  240.0     NaN   
...                      ...     ...  ...           ...    ...     ...   
119385                     5       2  ...    No Deposit  394.0     NaN   
119386                     5       3  ...    No Deposit    9.0     NaN   
119387                     5       2  ...    No Deposit    9.0     NaN   
119388                     5       2  ...    No Deposit   89.0     NaN   
119389                     7       2  ...    No Deposit    9.0     NaN   

       days_in_waiting_list customer_type     adr  \
0                         0     Transient    0.00   
1                         0     Transient    0.00   
2                         0     Transient   75.00   
3                         0     Transient   75.00   
4                         0     Transient   98.00   
...                     ...           ...     ...   
119385                    0     Transient   96.14   
119386                    0     Transient  225.43   
119387                    0     Transient  157.71   
119388                    0     Transient  104.40   
119389                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
0                                 0                          0   
1                                 0                          0   
2                                 0                          0   
3             

## Now we will try to check the shape of the Data

In [82]:
# Checking the shape of the data.
hotel_booking_df.shape  

(119390, 32)

## Now we wil see the various  columns and their names present in the data

In [83]:
# Checking the columns present in the data.
hotel_booking_df.columns   

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

## Since data type is important factor while data Wrangling, we check it and along with datatype we will have a note of the null values present in the dataset

In [84]:
# Checking the total non-null values count and datatypes of the coloumns.
hotel_booking_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

## Now we will understand the numerical information of our data.

In [85]:
# Now we will understand the numerical information of our data.
hotel_booking_df.describe(include ='all')

hotel    is_canceled      lead_time  arrival_date_year  \
count       119390  119390.000000  119390.000000      119390.000000   
unique           2            NaN            NaN                NaN   
top     City Hotel            NaN            NaN                NaN   
freq         79330            NaN            NaN                NaN   
mean           NaN       0.370416     104.011416        2016.156554   
std            NaN       0.482918     106.863097           0.707476   
min            NaN       0.000000       0.000000        2015.000000   
25%            NaN       0.000000      18.000000        2016.000000   
50%            NaN       0.000000      69.000000        2016.000000   
75%            NaN       1.000000     160.000000        2017.000000   
max            NaN       1.000000     737.000000        2017.000000   

       arrival_date_month  arrival_date_week_number  \
count              119390             119390.000000   
unique                 12                       NaN   
top                August                       NaN   
freq                13877                       NaN   
mean                  NaN                 27.165173   
std                   NaN                 13.605138   
min                   NaN                  1.000000   
25%                   NaN                 16.000000   
50%                   NaN                 28.000000   
75%                   NaN                 38.000000   
max                   NaN                 53.000000   

        arrival_date_day_of_month  stays_in_weekend_nights  \
count               119390.000000            119390.000000   
unique                        NaN                      NaN   
top                           NaN                      NaN   
freq                          NaN                      NaN   
mean                    15.798241                 0.927599   
std                      8.780829                 0.998613   
min                      1.000000                 0.000000   
25%                      8.000000                 0.000000   
50%                     16.000000                 1.000000   
75%                     23.000000                 2.000000   
max                     31.000000                19.000000   

        stays_in_week_nights         adults  ...  deposit_type          agent  \
count          119390.000000  119390.000000  ...        119390  103050.000000   
unique                   NaN            NaN  ...             3            NaN   
top                      NaN            NaN  ...    No Deposit            NaN   
freq                     NaN            NaN  ...        104641            NaN   
mean                2.500302       1.856403  ...           NaN      86.693382   
std                 1.908286       0.579261  ...           NaN     110.774548   
min                 0.000000       0.000000  ...           NaN       1.000000   
25%                 1.000000       2.000000  ...           NaN       9.000000   
50%                 2.000000       2.000000  ...           NaN      14.000000   
75%                 3.000000       2.000000  ...           NaN     229.000000   
max                50.000000      55.000000  ...           NaN     535.000000   

            company days_in_waiting_list customer_type            adr  \
count   6797.000000        119390.000000        119390  119390.000000   
unique          NaN                  NaN             4            NaN   
top             NaN                  NaN     Transient            NaN   
freq            NaN                  NaN         89613            NaN   
mean     189.266735             2.321149           NaN     101.831122   
std      131.655015            17.594721           NaN      50.535790   
min        6.000000             0.000000           NaN      -6.380000   
25%       62.000000             0.000000           NaN      69.290000   
50%      179.000000             0.000000           NaN      94.575000   
75%      270.000000             0.000000           

## By using (include = 'all') we can explore more about our coloums, such as its total no. of count, no. of unique values, frequency, etc.

## Next step we will take is to figue out the count of null values in different columns 

In [86]:
hotel_booking_df.isna().sum().sort_values(ascending = False)

company                           112593
agent                              16340
country                              488
children                               4
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
hotel                                  0
previous_cancellations                 0
days_in_waiting_list                   0
customer_type                          0
adr                                    0
required_car_parking_spaces            0
total_of_special_requests              0
reservation_status                     0
previous_bookings_not_canceled         0
is_repeated_guest                      0
is_canceled                            0
distribution_channel                   0
market_segment                         0
meal                                   0
babies                                 0
adults                                 0
stays_in_week_ni

## As we can clearly see that there are four coloumns containing the null values namely country, agent, company and children.


## Taking notes of the result will help us in the analysis the dataset and exploring the different correlation 

In [87]:
hotel_booking_exploring_table = [[column,hotel_booking_df[column].dtypes,hotel_booking_df[column].isnull().sum(),hotel_booking_df[column].nunique(),
                                   hotel_booking_df[column].unique()] for column in hotel_booking_df.columns] 
                                   
hotel_booking_df_exploring_table = pd.DataFrame(hotel_booking_exploring_table,columns = ['Column', 'Dtype', 'Null','nUnique', 'Unique'])

hotel_booking_df_exploring_table 


Column    Dtype    Null  nUnique  \
0                            hotel   object       0        2   
1                      is_canceled    int64       0        2   
2                        lead_time    int64       0      479   
3                arrival_date_year    int64       0        3   
4               arrival_date_month   object       0       12   
5         arrival_date_week_number    int64       0       53   
6        arrival_date_day_of_month    int64       0       31   
7          stays_in_weekend_nights    int64       0       17   
8             stays_in_week_nights    int64       0       35   
9                           adults    int64       0       14   
10                        children  float64       4        5   
11                          babies    int64       0        5   
12                            meal   object       0        5   
13                         country   object     488      177   
14                  market_segment   object       0        8   
15            distribution_channel   object       0        5   
16               is_repeated_guest    int64       0        2   
17          previous_cancellations    int64       0       15   
18  previous_bookings_not_canceled    int64       0       73   
19              reserved_room_type   object       0       10   
20              assigned_room_type   object       0       12   
21                 booking_changes    int64       0       21   
22                    deposit_type   object       0        3   
23                           agent  float64   16340      333   
24                         company  float64  112593      352   
25            days_in_waiting_list    int64       0      128   
26                   customer_type   object       0        4   
27                             adr  float64       0     8879   
28     required_car_parking_spaces    int64       0        5   
29       total_of_special_requests    int64       0        6   
30              reservation_status   object       0        3   
31         reservation_status_date   object       0      926   

                                               Unique  
0                          [Resort Hotel, City Hotel]  
1                                              [0, 1]  
2   [342, 737, 7, 13, 14, 0, 9, 85, 75, 23, 35, 68...  
3                                  [2015, 2016, 2017]  
4   [July, August, September, October, November, D...  
5   [27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 3...  
6   [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...  
7   [0, 1, 2, 4, 3, 6, 13, 8, 5, 7, 12, 9, 16, 18,...  
8   [0, 1, 2, 3, 4, 5, 10, 11, 8, 6, 7, 15, 9, 12,...  
9   [2, 1, 3, 4, 40, 26, 50, 27, 55, 0, 20, 6, 5, 10]  
10                    [0.0, 1.0, 2.0, 10.0, 3.0, nan]  
11                                   [0, 1, 2, 10, 9]  
12                        [BB, FB, HB, SC, Undefined]  
13  [PRT, GBR, USA, ESP, IRL, FRA, nan, ROU, NOR, ...  
14  [Direct, Corporate, Online TA, Offline TA/TO, ...  
15         [Direct, Corporate, TA/TO, Undefined, GDS]  
16                                             [0, 1]  
17  [0, 1, 2, 3, 26, 25, 14, 4, 24, 19, 5, 21, 6, ...  
18  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...  
19                     [C, A, D, E, G, F, H, L, P, B]  
20               [C, A, D, E, G, F, I, B, H, P, L, K]  
21  [3, 4, 0, 1, 2, 5, 17, 6, 8, 7, 10, 16, 9, 13,...  
22               [No Deposit, Refundable, Non Refund]  
23  [nan, 304.0, 240.0, 303.0, 15.0, 241.0, 8.0, 2...  
24  [nan, 110.0, 113.0, 270.0, 178.0, 240.0, 154.0...  
25  [0, 50, 47, 65, 122, 75, 101, 150, 125, 14, 60...  
26      [Transient, Contract, Transient-Party, Group]  
27  [0.0, 75.0, 98.0, 107.0, 103.0, 82.0, 105.5, 1...  
28                                    [0, 1, 2, 8, 3]  
29                                 [0, 1, 3, 2, 4, 5]  
30                     [Check-Out, Canceled, No-Show]  
31  [2015-07-01, 2015-07-02, 2015-07-03, 2015-05-0...

Data integrity is necessary to ensure a successful analysis
# In section we will take all the necessary measures to ensure that the data is clean and ready for analysis.

## we will create the copy of our original data

In [88]:
# we will create the copy of our original data
df_hotel_booking = hotel_booking_df.copy()

# As we know that there are four coloumns containing the null values namely country, agent, company and children. 
# We will replace or fill these null values 

Filling null values with 0 of company column.

In [89]:
# Filling null values with 0 of company column.
df_hotel_booking ['company'].fillna(0 , inplace = True) 

Filling null values with 0 of agent column.       

In [90]:
# Filling null values with 0 of agent column.          
df_hotel_booking['agent'].fillna(0 , inplace = True)  

Replacig the null values with others in country coloumn.

In [91]:
# Replacig the null values with 'others' in country coloumn.
df_hotel_booking['country'].fillna('Others',inplace = True)

Filling null values with its mean value of children column.

In [92]:
# Filling null values with its 'mean' of children column.
df_hotel_booking['children'].fillna(df_hotel_booking.children.mean(), inplace = True)

## In the datatype verification we saw that the numerical column agent, company and children have float64 datatype so we will convert it to int64 datatype

Converting company column datatype for float64 to int64

In [93]:
df_hotel_booking['company'] = df_hotel_booking['company'].astype(int) 

Converting children column datatype for float64 to int64

In [96]:
df_hotel_booking['children'] = df_hotel_booking['children'].astype(int) 

Converting agent column datatype for float64 to int64

In [97]:
df_hotel_booking['agent'] = df_hotel_booking['agent'].astype(int) 

# Now we will remove duplicate data which can affect the data analysis

In [98]:
df_hotel_booking[df_hotel_booking.duplicated()].shape

(31994, 32)

## We see that there are 31994 duplicate observations.

## so we will get ride of it as a data cleaning process.

In [99]:
df_hotel_booking.drop_duplicates(inplace = True)

## As we see that the sum of Adult,Babies, and Children are zero. This that it is not true data.
## So we will drop these observation as a part of data cleaning  

In [100]:
df_hotel_booking[df_hotel_booking.adults + df_hotel_booking.babies + df_hotel_booking.children == 0].shape

(166, 32)

In [101]:
df_hotel_booking.drop(df_hotel_booking[df_hotel_booking.adults + df_hotel_booking.babies + df_hotel_booking.children == 0].index, inplace = True)

# Now we see the final shape of our data

In [102]:
df_hotel_booking.shape

(87230, 32)

# We will rename columns to make the data set more understandable and and some new columns for better visualization

## Changing the adr column name to average_daily_rate

In [103]:
df_hotel_booking.rename(columns = {'adr' : 'df_hotel_booking'}, inplace = True)

# Adding new columns

## adding Total_group_members column for the total numbers of adults,children and babies

In [104]:
df_hotel_booking['Total_group_members'] = df_hotel_booking.adults + df_hotel_booking.children + df_hotel_booking.babies

## adding stays_in_weekend_and_weeks column for the total days of stay

In [105]:
df_hotel_booking['stays_in_weekend_and_weeks '] = df_hotel_booking.stays_in_weekend_nights + df_hotel_booking.stays_in_week_nights

In [106]:
df_hotel_booking.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 87230 entries, 0 to 119389
Data columns (total 34 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   hotel                           87230 non-null  object 
 1   is_canceled                     87230 non-null  int64  
 2   lead_time                       87230 non-null  int64  
 3   arrival_date_year               87230 non-null  int64  
 4   arrival_date_month              87230 non-null  object 
 5   arrival_date_week_number        87230 non-null  int64  
 6   arrival_date_day_of_month       87230 non-null  int64  
 7   stays_in_weekend_nights         87230 non-null  int64  
 8   stays_in_week_nights            87230 non-null  int64  
 9   adults                          87230 non-null  int64  
 10  children                        87230 non-null  int64  
 11  babies                          87230 non-null  int64  
 12  meal                           